src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -2864,30 +2864,49 @@ void executeCodeBlockActions(ImmutableArray<SyntaxNode> executableCodeBlocks, IE
                 first = false;
             }
 
-            var isSynthesizedRecordConstructor = semanticModel.IsSynthesizedRecordConstructor(declaredSymbol);
             Func<SyntaxNode, bool>? additionalFilter = semanticModel.GetSyntaxNodesToAnalyzeFilter(declaredNode, declaredSymbol);
-
-            bool shouldAddNode(SyntaxNode node) => (descendantDeclsToSkip == null || !descendantDeclsToSkip.Contains(node)) && (additionalFilter is null || additionalFilter(node));
             var nodeBuilder = ArrayBuilder<SyntaxNode>.GetInstance();
-            foreach (var node in declaredNode.DescendantNodesAndSelf(descendIntoChildren: shouldAddNode, descendIntoTrivia: true))
+            foreach (var node in declaredNode.DescendantNodesAndSelf(descendIntoChildren, descendIntoTrivia: true))
             {
-                if (!shouldAddNode(node))
+                // Skip nodes in descendantDeclsToSkip.
+                if (descendantDeclsToSkip?.Contains(node) == true)
                     continue;
 
-                // Skip node if we are doing partial analysis and the current node is outside the analysis scope.
-                if (isPartialDeclAnalysis && !analysisScope.ShouldAnalyze(node))
+                if (shouldSkipNodeAndDescendants(node))
```

Comment:
So, I wasn’t able to completely separate the filters into shouldAnalyzeNode and descendIntoChildren, but I simplified the descendIntoChildren to completely be based off descendantDeclsToSkip and the language specific filters can either filter out just an individual node OR a node and all it’s descendants (existing filter). This did simplify the code quite a bit. Let me know if this looks reasonable to you.

Reply:
> So, I wasn’t able to completely separate the filters into shouldAnalyzeNode and descendIntoChildren, but I simplified the descendIntoChildren to completely be based off descendantDeclsToSkip and the language specific filters can either filter out just an individual node OR a node and all it’s descendants (existing filter). This did simplify the code quite a bit. Let me know if this looks reasonable to you.

It feels like the new code is unnecessary verbose and probably a lot less optimal (nodes are filtered out later, more nodes are iterated through). Even if you find it difficult to pull apart what ```shouldAddNode``` used to be into two targeted delegates, I think it would be sufficient to add a new condition to the existing ```if``` in the loop. I am thinking about code like:
```
            bool shouldAddNode(SyntaxNode node) => (descendantDeclsToSkip == null || !descendantDeclsToSkip.Contains(node)) && (additionalFilter is null || additionalFilter(node));
            var nodeBuilder = ArrayBuilder<SyntaxNode>.GetInstance();
            foreach (var node in declaredNode.DescendantNodesAndSelf(descendIntoChildren: shouldAddNode, descendIntoTrivia: true))
            {
                if (shouldAddNode(node) &&
                    !semanticModel.ShouldSkipSyntaxNodeAnalysis(declaredNode, declaredSymbol) &&
                    (!isPartialDeclAnalysis || analysisScope.ShouldAnalyze(node)))
                {
                    nodeBuilder.Add(node);
                }
            }

```